gnovm/pkg/gnolang/ownership.go

```diff
@@ -376,3 +417,69 @@ func (tv *TypedValue) GetFirstObject(store Store) Object {
 		return nil
 	}
 }
+
+// GetFirstObjectAndAttach returns the value of [TypedValue.GetFirstObject], but
+// additionally sets its boundRealm
+func (tv *TypedValue) GetFirstObjectAndAttach(store Store) Object {
+	obj := tv.GetFirstObject(store)
+	if obj == nil {
+		// no object to set the bound realm to.
+		return nil
+	}
+
+	var path string
+	// infer original package using declared type
+	getPkgId := func(t Type) (pkgId PkgID) {
+		if dt, ok := t.(*DeclaredType); ok {
+			path = dt.GetPkgPath()
+			pkgId = PkgIDFromPkgPath(dt.GetPkgPath())
+		}
+		return
+	}
+
+	var boundRealm PkgID
+
+	switch cv := obj.(type) {
+	case *HeapItemValue:
+		boundRealm = getPkgId(cv.Value.T)
+	case *Block:
+		// assert to pointer value
+		if pv, ok := tv.V.(PointerValue); ok {
+			boundRealm = getPkgId(pv.TV.T)
+		}
+	case *BoundMethodValue:
+		// do nothing
+	case *MapValue, *StructValue, *ArrayValue:
+		// if it's a declared type, origin realm
+		// is deduced from type, otherwise zero.
+		boundRealm = getPkgId(tv.T)
+	default:
+		// do nothing
```

Comment:
panic, so future object types (if any) can be implemented and the panic help ensure to consider the logic here.